File: android/app/src/main/java/com/scholiast/android/domain/transcript/CueParser.kt
package com.scholiast.android.domain.transcript

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.longOrNull

/**
 * Parses YouTube caption track payloads into cues. Two input formats, mirroring
 * the desktop `video-transcript.ts`:
 *   1. **JSON3** (`baseUrl&fmt=json3`) — the format [TranscriptClient] always
 *      requests. Events with `tStartMs` start a cue; `aAppend` events (no
 *      `tStartMs`) accumulate their segs into the previous cue's text and never
 *      create a cue of their own — YouTube splits long lines into a start event
 *      plus one or more append events.
 *   2. **XML** — a faithful `parseCuesXml` port (srv3 `<p t="ms" d="ms">` with
 *      `<s>` word segments, then the simple `<text start="s" dur="s">` format),
 *      kept as a defensive fallback for when `fmt=json3` is ignored.
 *
 * All functions are pure (string in → cues out), so they are JVM-testable
 * without any Android dependency.
 */
object CueParser {

    private val json = Json { ignoreUnknownKeys = true }

    class ParseException(message: String, cause: Throwable? = null) : Exception(message, cause)

    /** Auto-detect: XML payloads start with `<`; JSON3 is an object. */
    fun parse(raw: String): List<TranscriptCue> =
        if (raw.trimStart().startsWith("<")) parseXml(raw) else parseJson3(raw)

    // --- JSON3 --------------------------------------------------------------

    /**
     * JSON3 shape (what `&fmt=json3` returns):
     * ```
     * { "wireMagic": "pb3", "events": [
     *     { "tStartMs": 0, "dDurationMs": 2680, "segs": [{"utf8": "Hello"}, {"tOffsetMs": 900, "acAsrConf": 0.9, "utf8": " world"}] },
     *     { "aAppend": 0, "segs": [{"utf8": ", and welcome."}] },
     *     ...
     * ]}
     * ```
     * A cue's `index` is its start event's position in `events` (append events
     * leave gaps). `endMs` = `tStartMs + dDurationMs` (no "+5 s" heuristic like
     * the desktop's `loadTranscript`, since real durations are available).
     */
    fun parseJson3(raw: String): List<TranscriptCue> {
        val events = try {
            (json.parseToJsonElement(raw) as? JsonObject)?.get("events") as? JsonArray
        } catch (e: Exception) {
            throw ParseException("malformed JSON3 caption payload", e)
        } ?: return emptyList()

        val cues = mutableListOf<TranscriptCue>()
        var startIndex = -1
        var startMs = -1L
        var endMs = -1L
        val text = StringBuilder()

        fun flush() {
            if (startIndex >= 0 && text.isNotBlank()) {
                cues += TranscriptCue(startIndex, startMs, endMs, text.toString().trim())
            }
            text.clear()
        }

        events.forEachIndexed { eventIndex, element ->
            val obj = element as? JsonObject ?: return@forEachIndexed
            val tStart = (obj["tStartMs"] as? JsonPrimitive)?.longOrNull
            if (tStart != null) {
                flush()
                startIndex = eventIndex
                startMs = tStart
                endMs = tStart + ((obj["dDurationMs"] as? JsonPrimitive)?.longOrNull ?: 0L)
                text.append(segsUtf8(obj))
            } else if (startIndex >= 0) {
                // aAppend (or an unknown non-start event): accumulate into the
                // previous cue; extend its end if a duration is given.
                text.append(segsUtf8(obj))
                val dur = (obj["dDurationMs"] as? JsonPrimitive)?.longOrNull
                if (dur != null && dur > 0) endMs = startMs + dur
            }
        }
        flush()
        return cues
    }

    private fun segsUtf8(obj: JsonObject): String =
        (obj["segs"] as? JsonArray)
            ?.mapNotNull { seg -> ((seg as? JsonObject)?.get("utf8") as? JsonPrimitive)?.contentOrNull }
            ?.joinToString("") ?: ""

    // --- XML (parseCuesXml port) ---------------------------------------------

    private val SRV3_P = Regex("""<p\s+t="(\d+)"(?:[^>]*?\sd="(\d+)")?[^>]*>([\s\S]*?)</p>""")
    private val SRV3_S = Regex("""<s[^>]*>([^<]*)</s>""")
    private val STRIP_TAGS = Regex("""<[^>]+>""")
    private val TEXT_TAG = Regex("""<text\s+start="([^"]*)"(?:[^>]*?\sdur="([^"]*)")?[^>]*>([\s\S]*?)</text>""")
    private val COLLAPSE_WS = Regex("""\s{2,}""")

    /** Port of the desktop `parseCuesXml`: srv3 first, then the simple format. */
    fun parseXml(xml: String): List<TranscriptCue> {
        parseSrv3(xml).let { if (it.isNotEmpty()) return it }
        return parseTextFormat(xml)
    }

    // srv3: <p t="ms" d="ms"><s>word</s>…</p> — one line = one cue.
    private fun parseSrv3(xml: String): List<TranscriptCue> {
        val cues = mutableListOf<TranscriptCue>()
        for (m in SRV3_P.findAll(xml)) {
            val startMs = m.groupValues[1].toLongOrNull() ?: continue
            val durMs = m.groupValues[2].toLongOrNull() ?: 0L
            val inner = m.groupValues[3]
            var text = ""
            for (s in SRV3_S.findAll(inner)) text += s.groupValues[1]
            if (text.isEmpty()) text = STRIP_TAGS.replace(inner, "")
            text = clean(text)
            if (text.isNotEmpty()) cues += TranscriptCue(cues.size, startMs, startMs + durMs, text)
        }
        return cues
    }

    // Simple format: <text start="s" dur="s">…</text> — timestamps in seconds.
    private fun parseTextFormat(xml: String): List<TranscriptCue> {
        val cues = mutableListOf<TranscriptCue>()
        for (m in TEXT_TAG.findAll(xml)) {
            val start = m.groupValues[1].toFloatOrNull() ?: 0f
            val dur = m.groupValues[2].toFloatOrNull() ?: 0f
            val text = clean(STRIP_TAGS.replace(m.groupValues[3], ""))
            if (text.isNotEmpty()) {
                cues += TranscriptCue(cues.size, (start * 1000).toLong(), ((start + dur) * 1000).toLong(), text)
            }
        }
        return cues
    }

    // Collapse whitespace first, then decode entities — same order as the TS.
    private fun clean(text: String): String =
        decodeEntities(text.replace('\n', ' ').replace(COLLAPSE_WS, " ")).trim()

    private fun decodeEntities(text: String): String {
        var s = text
            .replace("&amp;", "&").replace("&lt;", "<").replace("&gt;", ">")
            .replace("&quot;", "\"").replace("&#39;", "'").replace("&apos;", "'")
        s = HEX_ENTITY.replace(s) { m ->
            val cp = m.groupValues[1].toIntOrNull(16)
            if (cp != null && cp in 0..0x10FFFF) String(Character.toChars(cp)) else m.value
        }
        s = DEC_ENTITY.replace(s) { m ->
            val cp = m.groupValues[1].toIntOrNull()
            if (cp != null && cp in 0..0x10FFFF) String(Character.toChars(cp)) else m.value
        }
        return s
    }

    private val HEX_ENTITY = Regex("&#x([0-9a-fA-F]+);")
    private val DEC_ENTITY = Regex("&#(\\d+);")
}

File: scholiast_flutter/lib/core/algorithms/cue_parser.dart
import 'dart:convert';
import '../models/transcript_models.dart';

/// Parses YouTube caption track payloads into [Cue]s.
///
/// Supports two input formats mirroring the desktop `video-transcript.ts`:
/// 1. **JSON3** (`baseUrl&fmt=json3`): Events with `tStartMs` start a cue;
///    `aAppend` events (no `tStartMs`) accumulate their segments into the
///    previous cue's text and never create a cue of their own.
/// 2. **XML**: srv3 `<p t="ms" d="ms">` with `<s>` word segments, and the
///    simple `<text start="s" dur="s">` format.
class CueParser {
  const CueParser._();

  static final RegExp _srv3P =
      RegExp(r'<p\s+t="(\d+)"(?:[^>]*?\sd="(\d+)")?[^>]*>([\s\S]*?)</p>');
  static final RegExp _srv3S = RegExp(r'<s[^>]*>([^<]*)</s>');
  static final RegExp _stripTags = RegExp(r'<[^>]+>');
  static final RegExp _textTag = RegExp(
      r'<text\s+start="([^"]*)"(?:[^>]*?\sdur="([^"]*)")?[^>]*>([\s\S]*?)</text>');
  static final RegExp _collapseWs = RegExp(r'\s{2,}');
  static final RegExp _hexEntity = RegExp(r'&#x([0-9a-fA-F]+);');
  static final RegExp _decEntity = RegExp(r'&#(\d+);');

  /// Auto-detects format: XML payloads start with `<`; JSON3 is a JSON object.
  static List<Cue> parse(String raw) {
    if (raw.trimLeft().startsWith('<')) {
      return parseXml(raw);
    } else {
      return parseJson3(raw);
    }
  }

  // --- JSON3 ----------------------------------------------------------------

  /// Parses YouTube `&fmt=json3` caption payloads.
  ///
  /// A cue's [Cue.cueIndex] is its start event's position in `events`
  /// (`aAppend` events leave gaps).
  static List<Cue> parseJson3(String raw) {
    final dynamic decoded;
    try {
      decoded = jsonDecode(raw);
    } catch (e) {
      throw ParseException('malformed JSON3 caption payload', e);
    }

    if (decoded is! Map<String, dynamic>) {
      return const [];
    }

    final events = decoded['events'];
    if (events is! List) {
      return const [];
    }

    final cues = <Cue>[];
    var startIndex = -1;
    var startMs = -1.0;
    var endMs = -1.0;
    final textBuf = StringBuffer();

    void flush() {
      final textStr = textBuf.toString().trim();
      if (startIndex >= 0 && textStr.isNotEmpty) {
        final startSec = startMs / 1000.0;
        final durationSec = (endMs - startMs) / 1000.0;
        cues.add(Cue(
          start: startSec,
          duration: durationSec,
          text: textStr,
          cueIndex: startIndex,
        ));
      }
      textBuf.clear();
    }

    for (var eventIndex = 0; eventIndex < events.length; eventIndex++) {
      final element = events[eventIndex];
      if (element is! Map<String, dynamic>) continue;

      final tStart = element['tStartMs'];
      if (tStart is num) {
        flush();
        startIndex = eventIndex;
        startMs = tStart.toDouble();
        final dur = (element['dDurationMs'] as num?)?.toDouble() ?? 0.0;
        endMs = startMs + dur;
        textBuf.write(_segsUtf8(element));
      } else if (startIndex >= 0) {
        // aAppend (or unknown non-start event): accumulate into previous cue;
        // extend end if a duration is given.
        textBuf.write(_segsUtf8(element));
        final dur = (element['dDurationMs'] as num?)?.toDouble();
        if (dur != null && dur > 0) {
          endMs = startMs + dur;
        }
      }
    }
    flush();
    return cues;
  }

  static String _segsUtf8(Map<String, dynamic> obj) {
    final segs = obj['segs'];
    if (segs is! List) return '';
    final buf = StringBuffer();
    for (final seg in segs) {
      if (seg is Map<String, dynamic>) {
        final utf8 = seg['utf8'];
        if (utf8 != null) {
          buf.write(utf8.toString());
        }
      }
    }
    return buf.toString();
  }

  // --- XML (parseCuesXml port) -----------------------------------------------

  /// Parses XML caption formats: tries srv3 format first, then simple format.
  static List<Cue> parseXml(String xml) {
    final srv3 = _parseSrv3(xml);
    if (srv3.isNotEmpty) return srv3;
    return _parseTextFormat(xml);
  }

  // srv3: <p t="ms" d="ms"><s>word</s>…</p> — one line = one cue.
  static List<Cue> _parseSrv3(String xml) {
    final cues = <Cue>[];
    for (final m in _srv3P.allMatches(xml)) {
      final startMsStr = m.group(1);
      if (startMsStr == null) continue;
      final startMs = int.tryParse(startMsStr);
      if (startMs == null) continue;

      final durMsStr = m.group(2);
      final durMs = durMsStr != null ? (int.tryParse(durMsStr) ?? 0) : 0;

      final inner = m.group(3) ?? '';
      final sMatches = _srv3S.allMatches(inner).toList();
      var text = '';
      if (sMatches.isNotEmpty) {
        text = sMatches.map((s) => s.group(1) ?? '').join();
      }
      if (text.isEmpty) {
        text = inner.replaceAll(_stripTags, '');
      }
      text = clean(text);
      if (text.isNotEmpty) {
        cues.add(Cue(
          start: startMs / 1000.0,
          duration: durMs / 1000.0,
          text: text,
          cueIndex: cues.length,
        ));
      }
    }
    return cues;
  }

  // Simple format: <text start="s" dur="s">…</text> — timestamps in seconds.
  static List<Cue> _parseTextFormat(String xml) {
    final cues = <Cue>[];
    for (final m in _textTag.allMatches(xml)) {
      final startStr = m.group(1) ?? '0';
      final durStr = m.group(2);
      final start = double.tryParse(startStr) ?? 0.0;
      final dur = durStr != null ? (double.tryParse(durStr) ?? 0.0) : 0.0;
      final inner = m.group(3) ?? '';
      final text = clean(inner.replaceAll(_stripTags, ''));
      if (text.isNotEmpty) {
        cues.add(Cue(
          start: start,
          duration: dur,
          text: text,
          cueIndex: cues.length,
        ));
      }
    }
    return cues;
  }

  /// Collapses whitespace and decodes XML/HTML entities.
  static String clean(String text) {
    final collapsed =
        text.replaceAll('\n', ' ').replaceAll(_collapseWs, ' ');
    return decodeEntities(collapsed).trim();
  }

  /// Decodes named, decimal, and hex XML entities.
  static String decodeEntities(String text) {
    var s = text
        .replaceAll('&amp;', '&')
        .replaceAll('&lt;', '<')
        .replaceAll('&gt;', '>')
        .replaceAll('&quot;', '"')
        .replaceAll('&#39;', "'")
        .replaceAll('&apos;', "'");

    s = s.replaceAllMapped(_hexEntity, (m) {
      final hexStr = m.group(1);
      if (hexStr == null) return m.group(0)!;
      final cp = int.tryParse(hexStr, radix: 16);
      if (cp != null && cp >= 0 && cp <= 0x10FFFF) {
        return String.fromCharCode(cp);
      }
      return m.group(0)!;
    });

    s = s.replaceAllMapped(_decEntity, (m) {
      final decStr = m.group(1);
      if (decStr == null) return m.group(0)!;
      final cp = int.tryParse(decStr, radix: 10);
      if (cp != null && cp >= 0 && cp <= 0x10FFFF) {
        return String.fromCharCode(cp);
      }
      return m.group(0)!;
    });

    return s;
  }
}
